okio/jvm/src/main/java/okio/SelectRunner.kt

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import java.io.IOException
+import java.nio.channels.ClosedChannelException
+import java.nio.channels.SelectableChannel
+import java.nio.channels.SelectionKey
+import java.nio.channels.Selector
+import java.util.concurrent.ConcurrentLinkedQueue
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+suspend fun <T> SelectableChannel.selectAsync(ops: Int, block: () -> T): T {
+  val selectRunner = SelectRunner.instance
+  // TODO: cancel the key when the coroutine is canceled?
+  return suspendCoroutine { continuation ->
+    selectRunner.tasks.add(SelectRunner.Task(this, ops, block, continuation))
+    selectRunner.selector.wakeup()
+  }
+}
+
+internal class SelectRunner internal constructor() : Thread("Okio SelectRunner") {
+  init {
+    isDaemon = true
+  }
+
+  val selector = Selector.open()
+  val tasks = ConcurrentLinkedQueue<Task<*>>()
+
+  override fun run() {
+    while (true) {
+      val i = tasks.iterator()
+      while (i.hasNext()) {
+        val task = i.next()
+        if (task.call(selector)) {
+          i.remove()
+        }
+      }
+
+      // Register selection keys for everything recently added.
+      // TODO: enumerate the keys and decide which have timed out?
+      // TODO: sample the timeout so we know when to abort?
+
+      // Wait until something is ready.
+      selector.select()
+      selector.selectedKeys().clear()
+    }
+  }
+  internal class Task<T>(
+    private val channel: SelectableChannel,
+    private val ops: Int,
+    private val block: () -> T,
+    private val continuation: Continuation<T>
+  ) {
+    private var key: SelectionKey? = null
+
+    /** Returns true if this task is complete. */
+    fun call(selector: Selector): Boolean {
+      val key = key
+
+      // If the task hasn't been registered yet, do that first.
+      if (key == null) {
+        try {
+          this.key = channel.register(selector, ops, this)
+        } catch (e: ClosedChannelException) {
+          continuation.resumeWithException(e)
+          return true
+        }
+        return false
+      }
+
+      // If the channel has been closed, fail the task.
+      if (!key.isValid) {
+        continuation.resumeWithException(IOException("canceled"))
```

Comment:
`key.isValid` will also be false if the key has been canceled, right? I know you haven't implemented it yet but you have a todo above – if the key was canceled due to coroutine cancellation, this `resumeWithException` will just report the `IOException` to the uncaught exception handler. That's probably not what we want in this case, instead we should just not create the `IOException` at all and return true immediately.

If the key is invalid for some other reason (e.g. channel was closed, as you've commented), then we should still resume with exception.

There's at least two approaches to this:
1. Rely on the continuation itself as the source-of-truth for whether the key is invalid due to coroutine cancelation or not, or
2. Keep track explicitly ourselves.

The naïve way to implement 1 be just checking the active flag:
```kotlin
if (!key.isValid) {
  if (continuation.isActive) continuation.resumeWithException(IOException("canceled"))
  return true
}
```
But that creates a race where the `IOException` might get reported if the coroutine was canceled after the channel was closed. That might be ok since the key wasn't canceled due to coroutine cancelation, so technically we should still throw the exception. Another solution would be to use the atomic resume API:
```kotlin
if (!key.isValid) {
  continuation.tryResumeWithException(IOException("canceled"))?.let(continuation::completeResume)
  return true
}
```
But that API is marked internal so we that's just asking for trouble.

The safest solution is to do something like add an atomic boolean to `Task` that we can set to false in `invokeOnCancellation` before we cancel the key, and then check here instead of the `isActive` flag.

Reply:
Yep. We'll need to handle both coroutine cancellation and socket close.